File: leetcode_alg/solution20.cpp
//
// Created by wangheng on 2021/6/20.
// 有效括号问题
#include "iostream"
#include "stack"

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        if(s.length() == 0)
            return true;
        if(s.length() % 2 != 0)
            return false;
        stack<char> st;
        for(char c: s){
            switch (c) {
                case '(':{
                    st.push(c);
                    break;
                }
                case ')':{
                    if(st.size() == 0){
                        return false;
                    }else{
                        char &i = st.top();
                        if(i != '('){
                            return false;
                        }
                        st.pop();
                    }
                    break;
                }
                case '[':{
                    st.push(c);
                    break;
                }
                case ']':{
                    if(st.size() == 0){
                        return false;
                    }else{
                        char &i = st.top();
                        if(i != '['){
                            return false;
                        }
                        st.pop();
                    }
                    break;
                }
                case '{':{
                    st.push(c);
                    break;
                }
                case '}':{
                    if(st.size() == 0){
                        return false;
                    }else{
                        char &i = st.top();
                        if(i != '{'){
                            return false;
                        }
                        st.pop();
                    }
                    break;
                }
                default:{
                    continue;
                }
            }
        }
        if(st.size() == 0){
            return true;
        }
        return false;
    }
};

int main(){
    Solution *s = new Solution;
    cout<<s->isValid("()[]{}]")<<endl;
    return 0;
}

File: golang/solution/solution20.go
package solution

import (
	"fmt"

	"herywang.com/v2/stl"
)

var charMap = map[rune]rune{
	'(': ')',
	'[': ']',
	'{': '}',
}

func isValid(s string) bool {

	stack := &stl.Stack{}

	if len(s) == 0 {
		return true
	}
	if len(s)%2 != 0 {
		return false
	}

	for _, c := range s {
		if stack.IsEmpty() {
			stack.Push(c)
		} else if val, error := stack.Top(); error == nil {
			r, _ := val.(rune)
			if charMap[r] == c {
				stack.Pop()
			} else {
				stack.Push(c)
			}
		}
	}

	return stack.IsEmpty()
}

func Solution20() {
	fmt.Println(isValid("([{)}])"))
}
